### ISSUE TITLE
`cfg!($crate)`, `cfg!(crate)`, `cfg!(self)`, `cfg!(Self)`, and `cfg!(super)` should not be accepted

### ISSUE THREAD

**Issue by theemathas** [rust-lang/rust#146968]
I tried this code:

```rust
macro_rules! foo {
    () => {
        cfg!($crate)
    };
}

fn main() {
    let b = foo!();
    println!("{b}");
}
```

I expected a compile error. Instead, I got the following warning, but the code compiles fine and prints `false`.

```
warning: unexpected `cfg` condition name: `$crate`
 --> src/main.rs:3:14
  |
3 |         cfg!($crate)
  |              ^^^^^^
...
8 |     let b = foo!();
  |             ------ in this macro invocation
  |
  = help: expected names are: `docsrs`, `feature`, and `test` and 31 more
  = help: consider using a Cargo feature instead
  = help: or consider adding in `Cargo.toml` the `check-cfg` lint config for the lint:
           [lints.rust]
           unexpected_cfgs = { level = "warn", check-cfg = ['cfg($crate)'] }
  = help: or consider adding `println!("cargo::rustc-check-cfg=cfg($crate)");` to the top of the `build.rs`
  = note: see <https://doc.rust-lang.org/nightly/rustc/check-cfg/cargo-specifics.html> for more information about checking conditional configuration
  = note: `#[warn(unexpected_cfgs)]` on by default
  = note: this warning originates in the macro `foo` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Since it's impossible to set the value of this `cfg`, I think this should cause a compile error.

See also #146967 for more nonsense with `$crate`.

### Meta

Reproducible on the playground with version `1.92.0-nightly (2025-09-23 975e6c8fec280816d24f)`

**Comment by theemathas** [comment]
@danielhenrymantilla discovered that `cfg!(crate)`, `cfg!(self)`, and `cfg!(Self)` are also accepted, even though other keywords are not accepted. I think this also seems wrong.

**Comment by petrochenkov** [comment]
Syntactically, cfg predicates are parsed as paths, because they are parsed as meta-items.
E.g. something like `#[cfg(a::b)]` also parses, it's just then rejected by a semantic check "`cfg` predicate key must be an identifier".
`crate`, `self`, etc are all valid paths, but there's no semantic check that rejects them.
Other (non-path) keywords are different because they are rejected at parsing time.

**Comment by petrochenkov** [comment]
FWIW, on command line the validation seems to be even weaker:
```
rustc main.rs --cfg struct --cfg crate // ok
```

**Comment by estebank** [comment]
As an unrelated nitpick, we should also fix the wording of the `help`:

```
= help: expected names are: `docsrs`, `feature`, and `test` and 31 more
```

should be

```
= help: expected names are: `docsrs`, `feature`, `test` and 31 more
```

**Comment by 78589** [comment]
x
